### PR TITLE
feat: modexp precompile 384 bits

### LINKED ISSUE #1496: dev: modexp
as seen together teccak will repair modexp and make kakarot great again ♥️
```[tasklist]
### Tasks
- [ ] check modexp specification
- [ ] ignore kakarot modexp implementation and start over
- [ ] use new circuits?? on starknet to reimplement modexp soundly
- [ ] decide whether to implement modexp not for arbitrary length uints, but rather up to 256 bits
```

### PR DESCRIPTION
**Pull request by tekkac**
Time spent on this PR: 

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

```bash
Running 8 test(s) from src/
[PASS] evm::precompiles::modexp::tests::test_modexp_berlin_empty_input (gas: ~133)
[PASS] evm::precompiles::modexp::tests::test_modexp_eip198_example_1 (gas: ~635)
[PASS] evm::precompiles::modexp::tests::test_modexp_circuit (gas: ~544)
[PASS] evm::precompiles::modexp::tests::test_modexp_edge_case_circuit (gas: ~9)
[PASS] evm::precompiles::modexp::tests::test_modexp_eip198_example_2 (gas: ~312)
[PASS] evm::precompiles::modexp::tests::test_modexp_precompile_input_output_worst (gas: ~953)
[PASS] evm::precompiles::modexp::tests::test_modexp_circuit_worst_case (gas: ~953)
[PASS] evm::precompiles::modexp::tests::test_modexp_precompile_input_output_all_sizes (gas: ~22497)
```

Resolves #1496 

## What is the new behavior?

- modexp precompiled implemented with circuit up to 384 bits (48 bytes)
- passes basic tests and some EIP test vectors
- Custom input output needed because of non-standard integer types from circuits.

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1513)

### CODE REVIEW

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo
Comment (ClementWalter) [inline cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:1]: why are some tests commented?

Comment (tekkac) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:1]: Those tests won't pass as the scope of the issue is to impl modexp for up to 384 bits.
I kept the tests code for reference, let me know how best to handle it.

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo [hunk 1]
```diff
@@ -88,18 +87,158 @@ pub impl ModExp of Precompile {
         let gas = calc_gas(base_len.into(), exp_len.into(), mod_len.into(), exp_highp);
 
         // Padding is needed if the input does not contain all 3 values.
-        let base = input.slice_right_padded(0, base_len);
-        let exponent = input.slice_right_padded(base_len, exp_len);
-
         let (mod_start_idx, _) = base_len.overflowing_add(exp_len);
```
Comment (enitrat) [inline cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:90]: note (for us): this should probably not be overflowing_add here. if base_len + exp_len > u32 there's something wrong in the input (which will lead to an OOG prior to that anyway)

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo [hunk 2]
```diff
@@ -88,18 +87,158 @@ pub impl ModExp of Precompile {
...
+    let outputs = circuit_inputs.done().eval(modulus).unwrap();
+    (outputs.get_output(new_res), outputs.get_output(new_base))
+}
 
-        let return_data = output.pad_left_with_zeroes(mod_len);
-        Result::Ok((gas.into(), return_data))
+// Compute the modular exponentiation
+// x^y mod n
+pub fn modexp_circuit(x: u384, y: u384, n: u384) -> u384 {
+    println!("{:?} {:?} {:?}", x, y, n);
```
Comment (enitrat) [inline cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:178]:
```suggestion
```

**Review by enitrat — COMMENTED** [review]
great :)

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo [hunk 3]
```diff
@@ -88,18 +87,158 @@ pub impl ModExp of Precompile {
...
+    if x.is_zero() {
+        return 0.into();
+    }
```
[the remark this answers is not in the document]
Comment (obatirou) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:181]: It seems so, only if both base and modulo are 0 it should return 0.
https://github.com/ethereum/execution-specs/blob/3282ee5bb88bdd56aabf2461360277ab8eac4032/src/ethereum/cancun/vm/precompiled_contracts/modexp.py#L48

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo (hunk 1 above)
Comment (obatirou) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:90]: Should this be added as a comment in code ?

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo (hunk 2 above)
Comment (tekkac) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:178]: indeed, will remove

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo (hunk 3 above)
[the remark this answers is not in the document]
Comment (tekkac) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:181]: yup, correcting this and also a issue with circuit with modulus 1.

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo [hunk 4]
```diff
@@ -88,18 +87,158 @@ pub impl ModExp of Precompile {
...
+    if s_limb3 != 0 {
+        while bits.len() != 288 {
+            bits.append(0);
+        }
+    }
+    while s_limb3 != 0 {
+        let (q, r) = core::traits::DivRem::div_rem(s_limb3, 2);
+        bits.append(r.into());
+        s_limb3 = q;
+    };
```
[the remark this answers is not in the document]
Comment (tekkac) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:240]: I can try again, the main issue is that only TryInto felt252 is impl'd for u96.

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo (hunk 3 above)
[the remark this answers is not in the document]
Comment (tekkac) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo]: pushed a fix and a test for edge cases

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo (hunk 4 above)
[the remark this answers is not in the document]
Comment (tekkac) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:257]: Tried again without success, libfuncs are missing for u96

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo [hunk 5]
```diff
@@ -88,18 +87,158 @@ pub impl ModExp of Precompile {
...
+
+    let outputs = circuit_inputs.done().eval(modulus).unwrap();
+    (outputs.get_output(new_res), outputs.get_output(new_base))
+}
 
-        let return_data = output.pad_left_with_zeroes(mod_len);
-        Result::Ok((gas.into(), return_data))
+// Compute the modular exponentiation
+// x^y mod n
+pub fn modexp_circuit(x: u384, y: u384, n: u384) -> u384 {
```
[the remark this answers is not in the document]
Comment (tekkac) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:189]: Let me know if the docstring is correct and sufficiently explicit

[the remark this answers is not in the document]
Comment (obatirou) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:189]: Doesn't is miss the check `Assert :: (modulus - 1) * (modulus - 1) does not overflow base` ?

[the remark this answers is not in the document]
Comment (tekkac) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:189]: It's not needed in our case since the circuit already does all the operations with modular arithmetic.

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo
Comment (Eikix) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:1]: can delete

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo (hunk 1 above)
Comment (Eikix) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:90]: what to do?

**Comment by Eikix** [comment]
Why are ef-tests failing?

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo (hunk 1 above)
Comment (tekkac) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:90]: this isn't my code but happy to modify if needed

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo [hunk 6]
```diff
@@ -88,18 +87,158 @@ pub impl ModExp of Precompile {
...
+        let limb2_128: u128 = self.slice(12, 12).pad_left_with_zeroes(16).from_be_bytes().unwrap();
+        let limb1_128: u128 = self.slice(24, 12).pad_left_with_zeroes(16).from_be_bytes().unwrap();
+        let limb0_128: u128 = self.slice(36, 12).pad_left_with_zeroes(16).from_be_bytes().unwrap();
+        let limb0: u96 = Into::<_, felt252>::into(limb0_128).try_into().unwrap();
+        let limb1: u96 = Into::<_, felt252>::into(limb1_128).try_into().unwrap();
+        let limb2: u96 = Into::<_, felt252>::into(limb2_128).try_into().unwrap();
+        let limb3: u96 = Into::<_, felt252>::into(limb3_128).try_into().unwrap();
+        Option::Some(u384 { limb0, limb1, limb2, limb3 })
+    }
+}
```
[the remark this answers is not in the document]
Comment (tekkac) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:145]: u96 internal type is missing too many impls (same as other comment)

**Comment by tekkac** [comment]
> Why are ef-tests failing?

passing now after updating branch

**Comment by enitrat** [comment]
thanks, will do a second pass soon

File: cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+    # Construct input data
+    input_data = Bytes(
+        encoded_base_len + encoded_exp_len + encoded_modulus_len + base + exp + modulus
+    )
+
+    # Calculate result
+    result = modexp_384_bits(input_data)
+
+    if result:
```
Comment (obatirou) [inline cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py]: Why this if ?
`modexp_384_bits` should always return `gas, output`

File: cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+@given(
+    data=st.data(),
+    base_len=st.integers(min_value=0, max_value=256),
+    exp_len=st.integers(min_value=0, max_value=256),
+    modulus_len=st.integers(min_value=0, max_value=256),
+    base_bytes=st.binary(min_size=0, max_size=256),
+    exp_bytes=st.binary(min_size=0, max_size=256),
+    modulus_bytes=st.binary(min_size=0, max_size=256),
+)
+@settings(max_examples=25)
```
Comment (obatirou) [inline cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py]: Why 25 and not more ?
same for other functions

File: cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
+from typing import Dict, List, Tuple
```
Comment (obatirou) [inline cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py:1]: Very nice work here 👍

File: cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+
+if __name__ == "__main__":
+    # Run the Hypothesis test
+    generate_test_modexp_384_bits_random_inputs()
+    generate_test_modexp_384_bits_non_empty_outputs()
+
+    # Generate and save the Cairo tests
+    print(f"Generated {len(test_cases)} test cases")
+    with open("generated_modexp_tests.cairo", "w") as f:
```
Comment (obatirou) [inline cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py:216]: I must have missed something: those generated tests are not in this PR ?
In a previous version, they were added

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo [hunk 11]
```diff
@@ -151,152 +286,1003 @@ fn calculate_iteration_count(exp_length: u64, exp_highp: u256) -> u64 {
...
+            limb2: 0xffffffffffffffffffffffff,
+            limb3: 0xffffffffffffffffffffffff
+        };
 
-    // the tests are taken from
-    // [revm](https://github.com/bluealloy/revm/blob/0629883f5a40e913a5d9498fa37886348c858c70/crates/precompile/src/modexp.rs#L175)
+    #[test]
+    fn test_modexp_circuit() {
+        let TWO_31_M1: u384 = 2147483647.into();
+        let TWO_31_M2: u384 = 2147483646.into();
```
Comment (obatirou) [inline cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:330]: I would suggest to rename M1 and M2 by `MINUS_1` `MINUS_2` to be more explicit

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo [hunk 12]
```diff
@@ -151,152 +286,1003 @@ fn calculate_iteration_count(exp_length: u64, exp_highp: u256) -> u64 {
...
+        let expected_result = [].span();
+        let expected_gas = 200;
+        assert_eq!(result, expected_result);
+        assert_eq!(gas, expected_gas);
     }
 
+    // To test all input sizes, we use the fact that:
+    // For prime p, a^(p-1) mod p = 1, for all a. (Fermat's little theorem)
+    // Using prime_deltas we get a prime_i = 256^(i+2) + prime_deltas[i] of size i+2 bytes.
+    // We can then use p = prime_i and a = p - 2 and expect the result to be 1 with padding
```
Comment (obatirou) [inline cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:429]: What are prime deltas ?
This explaination is not clear to me could you explain in more details ?

File: cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py (hunk 10 above)
Comment (obatirou) [reply cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py:216]: Oh got it, they were added to modexp.cairo directly

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo [hunk 13]
```diff
@@ -57,69 +56,205 @@ pub impl ModExp of Precompile {
             }
         };
 
+        // We only support inputs up to 48 bytes. We can early return with an empty span if any size
+        // is larger than that.
+        if base_len > MAX_INPUT_BYTE_SIZE
+            || exp_len > MAX_INPUT_BYTE_SIZE
+            || mod_len > MAX_INPUT_BYTE_SIZE {
+            return Result::Err(EVMError::InvalidParameter('input greater than 48 bytes'));
```
Comment (obatirou) [inline cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:64]: It should be added somewhere in the documentation that this will raise for input > 48 bytes

Comment (enitrat) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:64]: yes PR on kakarot docs incoming

File: cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo (hunk 12 above)
Comment (enitrat) [reply cairo/kakarot-ssj/crates/evm/src/precompiles/modexp.cairo:429]: from what I understand:
- 256^(i+2) gives a number that's exactly i+2 bytes
- prime_deltas[i] is a small "adjustment" value to adjust the number to a prime number

it's actually a super clever test strategy 🫨 better than bruteforce with hypotheses 😓.

**Review by enitrat — DISMISSED** [review]

**Review by enitrat — APPROVED** [review]
perfect now 💯

File: cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py
Comment (ClementWalter) [inline cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py:1]: can do a follow up PR, but in a second thought we said that we could move these tests to end to end to make direct call to the precompile, instead of unit with generated data.

**Review by ClementWalter — APPROVED** [review]
I'm fine with merging as is if it's all ✅ and do follow up PRs for
- possible improvements cc @feltroidprime 
- refactoring tests to end-to-end directly cc @enitrat
